### PR TITLE
Remove the use of BrowserStack for the retrieval of the JWT token within the BE regression pack for Feature/vta 807

### PR DESCRIPTION
**Pull request by ZeypEr**
## Description
Using MSAL4J library to get tokens for API calls. By this we are removing BrowserStack communication with BE suite. 

Related issue: https://dvsa.atlassian.net/browse/VTA-807

### CODE REVIEW

File: src/main/java/util/TokenFromMSAL4J.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+public class TokenFromMSAL4J {
+    private static Loader loader = new LocalLoaderImpl();
+    private static String authority;
+    private static Set<String> scope;
+    private static String clientId;
+    private static String username;
+    private static String password;
+    public static  String token;
+    public static void main(String args[]) throws Exception {}
```
Comment (dhcrees) [inline src/main/java/util/TokenFromMSAL4J.java:17]: Remove Main method

File: src/main/java/util/TokenFromMSAL4J.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                .build();
+
+        //Get list of accounts from the application's token cache, and search them for the configured username
+        //getAccounts() will be empty on this first call, as accounts are added to the cache when acquiring a token
+        Set<IAccount> accountsInCache = pca.getAccounts().join();
+        IAccount account = getAccountByUsername(accountsInCache, username);
+
+        //Attempt to acquire token when user's account is not in the application's token cache
+        IAuthenticationResult result = acquireTokenUsernamePassword(pca, scope, account, username, password);
+//        System.out.println("Account username: " + result.account().username());
```
Comment (dhcrees) [inline src/main/java/util/TokenFromMSAL4J.java:32]: Remove all the system.out's

File: src/main/java/util/TokenFromMSAL4J.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        token=result.idToken();
+        accountsInCache = pca.getAccounts().join();
+        account = getAccountByUsername(accountsInCache, username);
+
+        //Attempt to acquire token again, now that the user's account and a token are in the application's token cache
+        result = acquireTokenUsernamePassword(pca, scope, account, username, password);
+//        System.out.println("Account username: " + result.account().username());
+//        System.out.println("Access token:     " + result.accessToken());
+//        System.out.println("Id token:         " + result.idToken());
+        token=result.idToken();
```
Comment (dhcrees) [inline src/main/java/util/TokenFromMSAL4J.java:45]: Remove this assignment, return token directly

File: src/main/java/util/TokenFromMSAL4J.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        return token;
+    }
+    private static IAuthenticationResult acquireTokenUsernamePassword(PublicClientApplication pca,
```
Comment (dhcrees) [inline src/main/java/util/TokenFromMSAL4J.java:48]: Add Dock Block

File: pom.xml [hunk 5]
```diff
@@ -106,6 +106,11 @@
             <artifactId>aws-java-sdk-cloudwatch</artifactId>
             <version>1.11.942</version>
         </dependency>
+        <dependency>
+            <groupId>com.microsoft.azure</groupId>
+            <artifactId>msal4j</artifactId>
+            <version>1.13.0</version>
```
Comment (dhcrees) [inline pom.xml:112]: Is this the latest version

File: src/main/java/steps/ActivitiesSteps.java [hunk 6]
```diff
@@ -201,18 +201,18 @@ public void insertRecordInDynamo(String json, String table, String primaryKey) {
...
-        return WebDriverBrowsertack.checkAtfEmail(testerName);
-    }
-
-    @Step
-    public void validateAtfEmailDetails(WebDriver driver, String testStationName, String testStationPNumber,
-                                        String testerName, String startDate, String startTime) {
-        WebDriverBrowsertack.checkAtfEmailDetails(driver, testStationName, testStationPNumber, testerName, startDate,
-                startTime);
-    }
+//
```
Comment (dhcrees) [inline src/main/java/steps/ActivitiesSteps.java:204]: Remove commented out code.

File: src/main/java/util/TokenFromMSAL4J.java (hunk 3 above)
Comment (ZeypEr) [reply src/main/java/util/TokenFromMSAL4J.java]: done

File: pom.xml (hunk 5 above)
Comment (ZeypEr) [reply pom.xml:112]: done

File: src/main/java/util/TokenFromMSAL4J.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
+package util;
+
+import com.microsoft.aad.msal4j.*;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Set;
+
+public class TokenFromMSAL4J {
```
Comment (bristol-mike-jones) [inline src/main/java/util/TokenFromMSAL4J.java:8]: Did you consider using a client creds flow and getting a JWT programmatically?  Do you need the JWT for anything other than signing API requests?

Using a password flow isn't recommended by MS - appreciate that this is a limited use case in that its all non prod.

Comment (dhcrees) [reply src/main/java/util/TokenFromMSAL4J.java:8]: Mike, We're re-writing the pack before long anyway, so as a temporary measure I'd like to get this merged in.

File: src/main/java/util/Loader.java [hunk 8]
```diff
@@ -38,4 +38,10 @@ public interface Loader {
 
     String getAppScope();
 
+    String getAuthority();
+
+    String getClientId();
+
+
```
Comment (IuliaDVSA) [inline src/main/java/util/Loader.java:45]: Please remove blank lines

File: pom.xml [hunk 9]
```diff
@@ -106,6 +106,11 @@
             <artifactId>aws-java-sdk-cloudwatch</artifactId>
             <version>1.11.942</version>
         </dependency>
+        <dependency>
+            <groupId>com.microsoft.azure</groupId>
+            <artifactId>msal4j</artifactId>
```
Comment (IuliaDVSA) [inline pom.xml:111]: What version does this need to be?

File: src/main/java/util/InstanceToken.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+    public static String getToken() {
+        // Handle other exceptions accordingly
+        try {
+            return TokenFromMSAL4J.createToken();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return null;}
+}
```
Comment (IuliaDVSA) [inline src/main/java/util/InstanceToken.java:14]: Please add line at end of file

**Review by IuliaDVSA — CHANGES_REQUESTED** [review]
some minor changes and query, can see you've removed the leftover main method.

**Review by IuliaDVSA — COMMENTED** [review]
Please rebase with the latest develop

**Pull request closed without merging**